src/meta/meta_ckpt.c

```diff
@@ -209,8 +210,32 @@ __ckpt_set(WT_SESSION_IMPL *session, const char *fname, const char *v, bool use_
     if (use_base && session->dhandle != NULL) {
         WT_ERR(__wt_scr_alloc(session, 0, &tmp));
         WT_ASSERT(session, strcmp(session->dhandle->name, fname) == 0);
+
+        /* Check the metadata is not corrupted. */
+        meta_base = session->dhandle->meta_base;
+        meta_base_length = meta_base == NULL ? 0 : strlen(meta_base);
+#ifdef HAVE_DIAGNOSTIC
+        if (session->dhandle->orig_meta_base_length != meta_base_length ||
+          (session->dhandle->orig_meta_base == NULL && meta_base != NULL) ||
+          (session->dhandle->orig_meta_base != NULL && meta_base == NULL) ||
+          (session->dhandle->orig_meta_base != NULL && meta_base != NULL &&
+            !WT_STREQ(session->dhandle->orig_meta_base, meta_base)))
+            WT_ERR_PANIC(session, WT_PANIC,
+              "Corrupted metadata. The original metadata length was %lu while the new one is %lu. "
+              "The original metadata inserted was %s and the current "
+              "metadata is now %s.",
+              session->dhandle->orig_meta_base_length, meta_base_length,
+              session->dhandle->orig_meta_base, meta_base);
+#endif
+        WT_ASSERT(session, session->dhandle->orig_meta_base_length == meta_base_length);
```

Comment:
In non diagnostic builds, this expands as UNUSED and then the warning "set but not used" is raised for meta_base_length.

Reply:
It can be a real failure. You can put this above the large `if` and remove that piece from the conditional. Then in `HAVE_DIAGNOSTIC` we only need to check the string content.
```
if (session->dhandle->... != meta_base_length)
    WT_ERR_PANIC(session, WT_PANIC, "Corrupted metadata...");